src/box/txn.c

```diff
@@ -1385,6 +1404,8 @@ box_txn_set_isolation(uint32_t level)
 	if (level == TXN_ISOLATION_DEFAULT)
 		level = txn_default_isolation;
 	txn->isolation = level;
+	if (level == TXN_ISOLATION_LINEARIZABLE)
+		return txn_make_linearizable(txn);
```

Comment:
I don't quite like that `box_txn_set_isolation` may know sleep - it looks like a simple setter. Maybe, we could postpone this until the first statement is executed (`txn_begin_stmt`), like we postpone creation of the transaction in engine (`txn_begin_in_engine`)?

Reply:
We'll have another problem then: select will yield. We decided this is not ok while reviewing the RFC.
`txn_set_isolation` is a part of `box.begin()` anyway (or IPROTO_BEGIN), so I don't see why it shouldn't yield.

Maybe factor `txn_make_linearizable` out of `txn_set_isolation` and place it in the end of `box.begin()` and IPROTO_BEGIN pr
ocessing? I think that `txn_set_isolation` and `txn_make_linearizable` should work atomically, but I can live with that.

Reply:
Yielding in `select` isn't good.

BTW, I just realized, we can't call `box_txn_set_isolation` via FFI anymore, because it may yield.

Reply:
I've asked @igormunkin, waiting for his response now.

Reply:
It's better not to use FFI for the yielding routines. You can use FFI, if the routine yields unconditionally: JIT engine just blacklists this trace and doesn't compile it later (so there is no need in FFI here except this is a more convenient way to call your routine). You shouldn't use FFI, if yield is made under condition: the trace can be compiled for the call with no yield made, but the compiled trace might be executed with the call yielding underneath, that breaks trace assumptions and hence trace semantics. For the latter case tarantool `panic` call occurs and tarantool instance stops with `exit(EXIT_FAILURE)`.

Reply:
@igormunkin, thanks for the clarification!
I've created a Lua/C wrapper for `box_txn_set_isolation`:

<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/misc.cc b/src/box/lua/misc.cc
index db566903b..2c2096a6b 100644
--- a/src/box/lua/misc.cc
+++ b/src/box/lua/misc.cc
@@ -41,6 +41,7 @@
 #include "box/tuple_format.h"
 #include "box/lua/tuple.h"
 #include "box/xrow.h"
+#include "box/txn.h"
 #include "mpstream/mpstream.h"
 
 static uint32_t CTID_STRUCT_TUPLE_FORMAT_PTR;
@@ -220,6 +221,21 @@ lbox_select(lua_State *L)
 
 /* }}} */
 
+/**
+ * Lua/C wrapper over box_txn_set_isolation. Is used in lua sources instead of
+ * a ffi call because box_txn_set_isolation yields occasionally.
+ */
+static int
+lbox_txn_set_isolation(struct lua_State *L)
+{
+	if (lua_gettop(L) != 1 || !lua_isnumber(L, 1))
+		return luaL_error(L, "Usage txn_set_isolation(level)");
+	uint32_t level = lua_tonumber(L, 1);
+	int rc = box_txn_set_isolation(level);
+	lua_pushnumber(L, rc);
+	return 1;
+}
+
 /** {{{ Utils to work with tuple_format. **/
 
 struct tuple_format *
@@ -333,6 +349,7 @@ box_lua_misc_init(struct lua_State *L)
 	static const struct luaL_Reg boxlib_internal[] = {
 		{"select", lbox_select},
 		{"new_tuple_format", lbox_tuple_format_new},
+		{"txn_set_isolation", lbox_txn_set_isolation},
 		{NULL, NULL}
 	};
 
diff --git a/src/box/lua/schema.lua b/src/box/lua/schema.lua
index 2eead2d3a..6b8ecc648 100644
--- a/src/box/lua/schema.lua
+++ b/src/box/lua/schema.lua
@@ -426,7 +426,7 @@ box.begin = function(options)
         assert(builtin.box_txn_set_timeout(timeout) == 0)
     end
     if txn_isolation and
-       builtin.box_txn_set_isolation(txn_isolation) ~= 0 then
+       internal.txn_set_isolation(txn_isolation) ~= 0 then
         box.rollback()
         box.error()
     end
```
</details>